sentry_sdk/metrics.py

```diff
@@ -0,0 +1,575 @@
+
+        # Given the new weight we consider whether we want to force flush.
+        self._consider_force_flush()
+
+    def kill(self):
+        # type: (...) -> None
+        if self._flusher is None:
+            return
+
+        self._running = False
+        self._flush_event.set()
+        self._flusher.join()
+        self._flusher = None
+
+    def flush(self):
+        # type: (...) -> None
+        self._force_flush = True
+        self._flush()
+
+    def _consider_force_flush(self):
+        # type: (...) -> None
+        # It's important to acquire a lock around this method, since it will touch shared data structures.
+        total_weight = len(self.buckets) + self._buckets_total_weight
+        if total_weight >= self.MAX_WEIGHT:
+            self._force_flush = True
+            self._flush_event.set()
+
+    def _emit(
+        self,
+        flushable_buckets,  # type: (Iterable[Tuple[int, Dict[BucketKey, Metric]]])
+    ):
+        # type: (...) -> Envelope
+        encoded_metrics = _encode_metrics(flushable_buckets)
+        metric_item = Item(payload=encoded_metrics, type="statsd")
+        envelope = Envelope(items=[metric_item])
+        self._capture_func(envelope)
+        return envelope
+
+    def _serialize_tags(
+        self, tags  # type: Optional[MetricTags]
+    ):
+        # type: (...) -> MetricTagsInternal
+        if not tags:
+            return ()
+
+        rv = []
+        for key, value in tags.items():
+            # If the value is a collection, we want to flatten it.
+            if isinstance(value, (list, tuple)):
+                for inner_value in value:
+                    if inner_value is not None:
+                        rv.append((key, str(inner_value)))
+            elif value is not None:
+                rv.append((key, str(value)))
+
+        # It's very important to sort the tags in order to obtain the
+        # same bucket key.
+        return tuple(sorted(rv))
+
+
+def _get_aggregator_and_update_tags(key, tags):
+    # type: (str, Optional[MetricTags]) -> Tuple[Optional[MetricsAggregator], Optional[MetricTags]]
+    """Returns the current metrics aggregator if there is one."""
+    hub = Hub.current
+    client = hub.client
+    if client is None or client.metrics_aggregator is None:
+        return None, tags
+
+    updated_tags = dict(tags or ())  # type: Dict[str, MetricTagValue]
+    updated_tags.setdefault("release", client.options["release"])
+    updated_tags.setdefault("environment", client.options["environment"])
+
+    scope = hub.scope
+    transaction_source = scope._transaction_info.get("source")
+    if transaction_source in GOOD_TRANSACTION_SOURCES:
+        transaction = scope._transaction
+        if transaction:
+            updated_tags.setdefault("transaction", transaction)
+
+    callback = client.options.get("_experiments", {}).get("before_emit_metric")
+    if callback is not None:
+        if not callback(key, updated_tags):
```

Comment:
If the callback doesn't return anything, this will now drop the metric. Is that intended or do you better want to screen for an explicit `False`? Since there's no other return value that would be used, this could be a pitfall.

Reply:
That's intentional and consistent with the behavior of `before_send`. When it returns `None` then that also discards the event.